File: java/src/main/java/ml/wonwoo/springdatashowcase/data/InitializerInvokerData.java
package ml.wonwoo.springdatashowcase.data;

import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.stereotype.Component;

import java.util.List;

@Component
public class InitializerInvokerData implements ApplicationRunner {

    private final List<InitializerData> initializerData;

    public InitializerInvokerData(List<InitializerData> initializerData) {
        this.initializerData = initializerData;
    }


    @Override
    public void run(ApplicationArguments args) {
        this.initializerData.forEach(InitializerData::initialize);
    }

    @Component
    static class NoOpDataInitializer implements InitializerData {

        @Override
        public void initialize() {

        }
    }
}

File: kotlin/src/main/kotlin/ml/wonwoo/springdatashowcase/data/InitializerInvokerData.kt
package ml.wonwoo.springdatashowcase.data

import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.stereotype.Component

@Component
class InitializerInvokerData(private val initializerData: List<InitializerData>) : ApplicationRunner {

    override fun run(args: ApplicationArguments?) {
        initializerData.forEach { it.initialize() }
    }

    @Component
    class NoopInitializerData : InitializerData {

        override fun initialize() {

        }
    }
}